Nmt/conv

noah-tutt-praetorian opened the pull request:
## Summary (required)

Briefly describe the changes and the issue it addresses. It helps the reviewers understand the context.

## JIRA (required)

- Add reference to the JIRA ticket. It helps connect JIRA with this PR.

peter-kwan commented on praetorian_cli/handlers/utils.py:
```diff
@@ -44,6 +44,11 @@ class Risk(Enum):
+    CLOSED_LOW_FALSE_POSITIVE = "CLF"
+    CLOSED_MEDIUM_FALSE_POSITIVE = "CMF"
+    CLOSED_HIGH_FALSE_POSITIVE = "CHF"
+    CLOSED_CRITICAL_FALSE_POSITIVE = "CCF"
```
Could you add the out of scope statuses?

peter-kwan commented:
1 change. Otherwise good.

@Ayushi-12  -- once you added the status on out of scope, you can go ahead merge and publish a new version.

peter-kwan commented:
Please also update https://github.com/praetorian-inc/praetorian-cli/blob/main/docs/terminology.md

noah-tutt-praetorian committed:
feat: add conversation CLI with Rich TUI interface

Implement interactive conversation CLI following existing aegis TUI patterns:

- Add 'conversation' command to main CLI with login requirement
- Create Rich-based TUI with conversation loop and message history
- Support conversation commands: help, clear, new, quit
- Format AI responses with panels and JSON syntax highlighting
- Auto-generate conversation IDs and maintain session state
- Integration with Chariot conversation API endpoint

Usage: praetorian-cli conversation

Features:
- Clean command-driven interface matching aegis style
- Formatted display of query results with JSON highlighting
- Session management with conversation persistence
- Help system with query examples
- Keyboard interrupt handling and graceful exit

noah-tutt-praetorian committed:
fix: update conversation CLI to use correct SDK methods

- Fix decorator import: use cli_handler instead of with_chariot_login_required
- Fix API call: use sdk._make_request and sdk.url() methods
- Add test script for debugging conversation API integration

Ready for testing with corrected Bedrock model ID.

noah-tutt-praetorian committed:
cleanup: remove test script

Test debugging complete - CLI implementation verified working.

noah-tutt-praetorian committed:
debug: add parameter parsing flexibility and logging

- Add debug logging to see actual tool input parameter keys
- Handle both nested query format and direct parameter format
- Add getKeys helper for debugging tool input structure

This will help identify how the AI is structuring tool calls
and enable flexible parameter parsing.

noah-tutt-praetorian committed:
feat: add tool output truncation in CLI display

Implement smart tool output display:

1. Truncate JSON tool output to first 5 lines for readability
2. Show truncation indicator with line count
3. Display AI summary prominently above tool output
4. Use color coding: green for complete, yellow for truncated
5. Maintain full JSON formatting for short outputs

This provides clean, digestible output while indicating when
full tool results are available but truncated for readability.

noah-tutt-praetorian committed:
docs: add comprehensive CLI testing guide

Add CONVERSATION_TEST.md with:
- Manual testing instructions for interactive CLI
- Expected output examples with summarization + truncation
- Alternative testing methods and troubleshooting
- Command examples for validation

Ready for manual testing of complete conversation flow.

noah-tutt-praetorian committed:
refactor: move conversation to agent subcommand

Move conversation from top-level to agent subcommand following MCP pattern:

- Add conversation as @agent.command() in handlers/agent.py
- Remove standalone conversation handler
- Update main.py imports and command registration
- Update test documentation for new command path

New usage:
  praetorian-cli chariot agent conversation

This follows the established pattern where AI features are
organized under the agent subcommand alongside MCP.

noah-tutt-praetorian committed:
feat: update CLI to use /planner endpoint

noah-tutt-praetorian committed:
feat: add conversation job display to CLI TUI header

noah-tutt-praetorian committed:
feat: add planner-output role support to conversation interface

- Add special display formatting for scan completion messages
- Support **Scan Complete** and **Scan Failed** message prefixes
- Green/red color coding for scan result status
- Enhanced UI feedback for asynchronous job completion events

noah-tutt-praetorian committed:
fix: implement server-controlled conversation creation in CLI

- Remove client-side conversation ID generation
- Send first message without conversationId field
- Extract conversation UUID from server response
- Use server-provided ID for subsequent messages in conversation
- Implement secure conversation creation flow (server controls IDs)

noah-tutt-praetorian committed:
fix: handle 201 status code for conversation creation

- Accept both 200 and 201 status codes as success
- Handle conversation creation response (201) properly
- Extract conversation UUID from creation response
- Show success message for conversation creation instead of error

noah-tutt-praetorian committed:
fix: simplify CLI response handling for unified 200 status

- Only handle 200 status code since backend now returns 200 for all success
- Extract conversation ID from first message response
- Simplify response parsing for unified conversation creation flow

noah-tutt-praetorian committed:
fix: handle string responses from planner API

- Add string response type handling for direct AI responses
- Handle both string and object response formats
- Fix 'str' object has no attribute 'get' error
- Support unified response format from conversation API

noah-tutt-praetorian committed:
feat: implement comprehensive conversation management and polling

- Add startup menu to choose new conversation or resume existing
- Implement conversation polling for real-time AI responses
- Add conversation listing with message counts and creation dates
- Replace immediate response with polling until AI responds
- Remove legacy response handling and queued message display
- Add conversation resume functionality with table selection
- Implement proper message polling and display pipeline

noah-tutt-praetorian committed:
fix: use existing /my endpoint for conversation and message listing

- Replace non-existent endpoints with sdk.my() calls
- Use #conversation# prefix for listing conversations
- Use #message#{conversation_id} prefix for message polling
- Fix conversation resume functionality with proper SDK integration
- Sort messages by timestamp for correct display order

noah-tutt-praetorian committed:
fix: use correct SDK search.by_key_prefix() method

- Replace incorrect my() usage with search.by_key_prefix()
- Use proper SDK search interface for conversations and messages
- Fix conversation resume and message polling functionality
- Follow established SDK patterns from existing codebase

noah-tutt-praetorian committed:
fix: implement continuous polling until AI responds

- Poll conversation messages continuously until AI (chariot role) responds
- Keep thinking spinner active while most recent message is from user
- Stop spinner only when AI provides response
- Handle asynchronous processing properly with real-time polling

noah-tutt-praetorian committed:
feat: add topic field to conversations and improve database permissions

Backend:
- Replace AmazonDynamoDBFullAccess with DatabaseRW policy for PlannerEventsFunction
- Add Topic field to Conversation model for first message preview
- Set conversation topic from first message (256 chars) on creation

CLI:
- Display conversation topic in resume conversation table
- Add topic column to conversation selection interface
- Truncate topics to 40 chars for table display

noah-tutt-praetorian committed:
feat: display tool execution progress in conversation interface

- Show tool call messages as '🔧 Executing {tool_name} tool...'
- Show tool response completion as '✅ Tool execution completed'
- Extract tool name from toolUseContent JSON for better UX
- Provide real-time feedback during tool chaining workflows

noah-tutt-praetorian committed:
debug: add logging to tool name extraction

- Add debug output to see actual toolUseContent structure
- Try multiple possible field names for tool name
- Help identify why tool names show as 'unknown'

noah-tutt-praetorian committed:
simplify: show generic tool execution messages without tool names

- Change to simple '🔧 Executing tool...' message
- Remove complex tool name extraction logic
- Clean up tool execution display for better simplicity

noah-tutt-praetorian committed:
cleanup: remove unneeded test and documentation files

- Remove CONVERSATION_TEST.md (testing documentation)
- Remove docs/webpage-linking.md (feature documentation)
- Remove test_tool_execution.py (test script)
- Keep only essential conversation UI implementation

noah-tutt-praetorian committed:
feat: add user query parameter support to SDK and CLI

- Add USER_FLAG to SDK globals following GLOBAL_FLAG pattern
- Add user parameter to by_term and by_key_prefix methods
- Update conversation and message queries in CLI to use user=True
- Enable user-scoped queries for conversations and messages in TUI
- Follow existing global parameter implementation pattern

noah-tutt-praetorian committed:
feat: implement continuous background polling for async job updates

- Add background polling thread that runs continuously
- Poll every 3 seconds for new messages even when not actively waiting
- Display job completion summaries immediately when they arrive
- Handle async scan results with special formatting and job update notifications
- Enable real-time job completion feedback without user interaction
- Clean thread shutdown on exit

noah-tutt-praetorian committed:
fix: correct Python syntax error in try/except/finally block

- Fix nested try/except structure that caused syntax error
- Properly structure exception handling for main conversation loop
- Ensure background polling cleanup happens on exit

noah-tutt-praetorian committed:
fix: use search.by_term for conversation job filtering

- Replace jobs.list(prefix_filter) with search.by_term for conversation job queries
- Use conversation:<uuid> search term with user=True parameter
- Fix job listing for proper conversation-scoped job filtering
- Enable correct job status display in conversation UI

noah-tutt-praetorian committed:
feat: enhance job display and add conversation history loading

Job Display Improvements:
- Separate queued (JQ) from running (JR) jobs in status counts
- Extract readable target names from asset keys (#asset#domain#hostname)
- Map job status codes to human text (JQ->Queued, JR->Running, etc.)
- Use proper blue color for queued jobs vs yellow for running

Conversation Resume:
- Load and display previous messages when resuming conversation
- Show complete conversation history with proper role formatting
- Include tool execution history and job completion summaries
- Set proper message count for background polling continuation

Enhanced UX with accurate job status reporting and full conversation context.

noah-tutt-praetorian committed:
debug: add target structure debugging to identify job target format

- Add debug output to understand actual job target structure
- Try multiple extraction approaches for target display
- Debug both target field and job key field contents
- Enable proper target name extraction once structure is understood

noah-tutt-praetorian committed:
fix: extract target from job key structure properly

- Extract target from job key format: #job#target#capability
- Parse nested asset keys within job target for readable display
- Use DNS field as fallback for target display
- Handle both direct targets and nested asset key targets
- Should now show proper hostnames instead of 'unknown'

noah-tutt-praetorian committed:
perf: remove message count from conversation resume menu

- Remove Messages column from conversation selection table
- Remove get_conversation_message_count method to eliminate API calls
- Speed up conversation resume menu loading
- Simplify conversation table to ID, Topic, and Created columns only

noah-tutt-praetorian committed:
refactor: implement unified efficient message loading with greater-than queries

- Consolidate to single background polling thread for all message loading
- Use efficient key:> search queries to load only new messages since last poll
- Replace message count tracking with message key tracking for precision
- Apply same efficient logic for both conversation history and real-time updates
- Improve performance by avoiding full conversation scans on each poll
- Enable true real-time updates with minimal API overhead

noah-tutt-praetorian committed:
feat: add mode switching to CLI with query as default

- Default to query mode for safe data discovery
- Add mode display in header with color coding (blue=query, green=agent)
- Add mode switching commands: 'query' and 'agent'
- Include mode parameter in API requests
- Update help documentation with mode descriptions
- Enable real-time mode switching within conversations

noah-tutt-praetorian committed:
update

noah-tutt-praetorian committed:
update
